src/EditorFeatures/Core/Implementation/Formatting/FormatCommandHandler.FormatDocument.cs

```diff
@@ -34,15 +60,41 @@ private bool TryExecuteCommand(FormatDocumentCommandArgs args, CommandExecutionC
                 return false;
             }
 
-            var formattingService = document.GetLanguageService<IEditorFormattingService>();
-            if (formattingService == null || !formattingService.SupportsFormatDocument)
+            if (!document.Project.Solution.Workspace.Options.GetOption(FeatureOnOffOptions.IsCodeCleanupRulesConfigured, LanguageNames.CSharp))
             {
-                return false;
+                ShowGoldBarForCodeCleanupConfiguration(document.Project.Solution.Workspace);
+                Format(args.TextView, document, null, context.WaitContext.UserCancellationToken);
             }
-
-            using (context.WaitContext.AddScope(allowCancellation: true, EditorFeaturesResources.Formatting_document))
+            else
             {
-                Format(args.TextView, document, null, context.WaitContext.UserCancellationToken);
+
+                using (context.WaitContext.AddScope(allowCancellation: true, EditorFeaturesResources.Formatting_document))
+                {
+                    var cancellationToken = context.WaitContext.UserCancellationToken;
+
+                    using (var transaction = new CaretPreservingEditTransaction(
+                        EditorFeaturesResources.Formatting, args.TextView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                    {
+                        var oldDocument = document;
+                        var codeFixChanges = _codeCleanupService.GetChangesForCleanupDocument(document, cancellationToken).Result.ToList();
```

Comment:
1.  what happens for languages like F#/TS here?
2. Calling .Result on this task can throw bad exceptions (for example, in cases like cancellation).

Reply:
Fixed